### ISSUE TITLE
CSS @import not behaving as documented

### ISSUE THREAD

**Issue by chalin** [dart-lang/angular2#39]
**Update**: this is still an issue with 3.0.0. I've updated the text below and added a link to sources that can be used to reproduce the issue.

---
The Dart equivalent of the [Dev Guide Component Styles example](https://angular.io/docs/ts/latest/guide/component-styles.html), doesn't behave like its TS counterpart when it comes to CSS `@import`s.

**Steps to reproduce and a minimal demo of the problem**

Get the [component-styles](https://github.com/angular-examples/component-styles) example project. Relevant files are:

- `lib/hero_details_componet.dart`:
> ``` dart
> @Component(
>     selector: 'hero-details',
>     template: ...,
>     styleUrls: const ['hero_details_component.css'],   /*<<<<<<<*/
>     directives: const [HeroTeamComponent])
> class HeroDetailsComponent {
>   @Input() Hero hero;
> }
> ```
- `lib/hero_details_component.css` (excerpt):
> ``` css
> /* Causes an ERROR:*/ @import 'hero_details_box.css';  /*<<<<<<<*/
> :host {
>   display: block;
>   border: 1px solid black;
> }
> ...
> ```
- `lib/hero_details_box.css`
> ```css
> :host { padding: 10px; }
> h2::after { content: " (from imported CSS)"; }
> ```

**Current behavior**

The `@import` line in `lib/hero_details_component.css` is currently commented out. If you uncomment the line (and remove the space between the `@` and the `import` tokens), and do a `pub serve` or `pub build` then you'll get an error:

``` nocode
Building component_styles...
Build error:
Transform StylesheetCompiler on component_styles|lib/hero_details_component.css threw error: Url component_styles|lib/hero_details_component.css.dart is not a valid asset: url
```

Replacing the `@import` argument with any relative path results in the same error.

If you copy `lib/hero_details_box.css` into the project's `web` folder and change the `@import` argument to an absolute path like:

> `@import '/hero_details_box.css';`

then the normal CSS selectors work, but `:host` styling doesn't.

**Expected/desired behavior**

According to the [Dev Guide Component Styles, CSS @imports subsection](https://angular.io/docs/ts/latest/guide/component-styles.html#css-imports):

> We can also import CSS files into our CSS files by using the standard CSS @import rule.
> In this case the URL is relative to the CSS file into which we are importing.

**Other information**

This was tested with **beta.17**. Possibly related (but different) issues:
- CSS @import with root path "/" bug https://github.com/angular/angular/issues/6449
- Transformer does not work well with web urls https://github.com/angular/angular/issues/2607
- styleUrls does not work with root path "/" https://github.com/angular/angular/issues/4974

Migrated from https://github.com/angular/angular/issues/8518.

**Comment by chalin** [comment]
cc @kwalrath

**Comment by ferhatb** [comment]
No need for @import in Dart. Dart has package resolution.

styleUrls: const ['package:somepackage/hero_details_component.css']

**Comment by chalin** [comment]
I think that there might have been a misunderstanding. You wrote:
> styleUrls: const ['package:somepackage/hero_details_component.css']

This issue is not about the interpretation of URLs to style files as they appear in Dart code in, say, `styleUrls` lists.

This issue is about how the [CSS @import rule](https://developer.mozilla.org/en-US/docs/Web/CSS/@import) URL argument gets interpreted. There seems to be an issue with the **`StylesheetCompiler` transformer**. `@import` is a CSS feature which is independent of TS or Dart.

**Comment by koorgoo** [comment]
@ferhatb @chalin This issue is closed by mistake.
Please, see the problem context from the documentation in [Component Styles](https://angular.io/docs/dart/latest/guide/component-styles.html#!#css-imports).

**Comment by kwalrath** [comment]
Reopening, since this is about CSS imports, not Dart imports.

**Comment by chalin** [comment]
@matanlurey, @leonsenft et all: this still doesn't work under 3.0.0. I've updated the issue text above and added a link to sources that can be used to reproduce the issue. Let me know what other info you might need.

**Comment by leonsenft** [comment]
Thanks @chalin, it looks like you've provided plenty of info. I'll look into this.

**Comment by leonsenft** [comment]
> If you copy `lib/hero_details_box.css` into the project's `web` folder and change the `@import` argument to an absolute path like:
>
> `@import '/hero_details_box.css';`
>
> then the normal CSS selectors work, but `:host` styling doesn't.

The compiler ignores `@import` statements for absolute URLs, so this effectively bypasses Angular's built-in `@import` support. The `@import` statement remains in the final CSS and ends up being handled by the browser's implementation. This is why it appears to work if the CSS file is moved to `web/`, because it's now at the location the browser requests. This is also why `:host` styling in these files ceases to work, because it's a feature shimmed by the compiler.

I think this might actually be intentional for use under _web/_ where resources aren't typically shimmed (since they're not associated with a component).

**Comment by chalin** [comment]
I'm still getting an error when I try this over the original example using Angular 3.1.0:
```
component-styles > pub build
Loading source assets... 
Loading angular2/transform/codegen and dart_to_js_script_rewriter transformers... 
Loading angular2 transformers... (6.4s)
Building component_styles... 
Build error:
Transform StylesheetCompiler on component_styles|lib/src/hero_details_component.css threw error: Url component_styles|lib/src/hero_details_component.css.dart is not a valid asset: url
package:angular2/src/compiler/output/path_util.dart 42:5                    _AssetUrl.parse
package:angular2/src/compiler/output/path_util.dart 14:29                   getImportModulePath
package:angular2/src/compiler/output/dart_emitter.dart 49:27                DartEmitter.emitStatements.<fn>
dart:collection-patch/compact_hash.dart 356                                 _HashVMBase&MapMixin&&_LinkedHashMapMixin.forEach
package:angular2/src/compiler/output/dart_emitter.dart 48:35                DartEmitter.emitStatements
package:angular2/src/compiler/offline_compiler.dart 159:40                  OfflineCompiler._createSourceModule
package:angular2/src/compiler/offline_compiler.dart 125:7                   OfflineCompiler.compileStylesheet
package:angular2/src/transform/stylesheet_compiler/processor.dart 32:26     processStylesheet.<fn>
```

**Comment by leonsenft** [comment]
@chalin I don't believe the fix made it into 3.1.0. You'll have to update to 4.0.0-alpha.

**Comment by chalin** [comment]
Ok, thanks for the clarification. I'll add that to my list of things to check under 4.0.
